xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiServletURLFactory.java

```diff
@@ -444,23 +446,53 @@ public URL createExternalURL(String spaces, String name, String action, String q
         return this.createURL(spaces, name, action, querystring, anchor, xwikidb, context);
     }
 
+    private URL buildURL(URL serverUrl, String path, Map<String, Object> queryParameters, XWikiContext context)
+    {
+        try {
+            URIBuilder uriBuilder = new URIBuilder(serverUrl.toURI()).setPath(path);
+            for (Map.Entry<String, Object> queryParameter : queryParameters.entrySet()) {
+                String key = queryParameter.getKey();
+                Object paramValue = queryParameter.getValue();
+
+                if (paramValue instanceof String) {
+                    uriBuilder.addParameter(key, (String) paramValue);
+                } else if (paramValue instanceof String[]) {
+                    for (String paramValueElement : (String[]) paramValue) {
+                        uriBuilder.addParameter(key, paramValueElement);
+                    }
+                } else if (paramValue instanceof Collection) {
+                    for (Object paramValueElement : (Collection) paramValue) {
+                        uriBuilder.addParameter(key, paramValueElement.toString());
+                    }
+                } else {
+                    uriBuilder.addParameter(key, paramValue.toString());
+                }
```

Comment:
So finally URIBuilder does not provide a simple API for that: I only handle collection, since I'm not sure how we should handle it for a map.